### DISCUSSION TITLE
deploy Winform-application: Could not load file or assembly 'CefSharp.Core.Runtime.dll

### DISCUSSION THREAD

**Discussion by mrasoftGithub** [cefsharp/CefSharp#3402] [Q&A]
Dear Team,

I want to deploy a C# Winform-application with CEFSharp. After publishing, a **setup.exe** is created but I got this error on clicking the **setup.exe**: **Could not load file or assembly 'CefSharp.Core.Runtime.dll'.**  

The program runs without problems in Visual Studio 2017 in the development-debug mode. Both '**CefSharp.Core.Runtime.dll**' and '**CefSharp.Core.dll**' are found in the deployment package.

I already read the other discussions and i tried accordingly but I still got the error. 

Please assist. Thanks,

Depending on the Platform Target architecture (x64, x86) and the build (debug , release) I added this property to the project file: `<CefSharpBuildAction>Content</CefSharpBuildAction>` 

For example I added this for a Debug x64 build platform: 
`<PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">`
`<DebugSymbols>true</DebugSymbols>`
`<OutputPath>bin\x64\Debug\</OutputPath>`
`<CefSharpBuildAction>Content</CefSharpBuildAction>`

the property is not supported by **AnyCPU** at the time of this writing.

-----
**The error does not seem to appear in Visual Studio 2019.**
-----

-----
**Solved by adding this to the project file in Visual Studio 2017:**
----

**Adding this solved the problem in which duplicates are removed from the manifest file**:
`<Target Name="CefSharpClickOnceVS2017" BeforeTargets="GenerateApplicationManifest">`
`	<ItemGroup>`
`		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.dll" />`
`		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.pdb" />`
`		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.xml" />`
`	</ItemGroup>`
`</Target>`

**Or adding this to the project file helped as well:**

`<ItemGroup>`
`    <Content Include="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\**\*" Exclude="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\x64\**\*;$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\locales\**\*.pak">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`      <Visible>false</Visible>`
`    </Content>`
`    <Content Include="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\**\en-US.*">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`    </Content>`
`    <Content Include="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\x64\**\*">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`      <Visible>false</Visible>`
`    </Content>`
`    <Content Include="$(SolutionDir)packages\CefSharp.Common.87.1.132\CefSharp\x64\**\CefSharp.BrowserSubprocess.*">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`      <Visible>false</Visible>`
`    </Content>`
`</ItemGroup>`

**What is used**:

- Visual Studio Community 2017 - Version 15.9.31
- Windows Forms App (.NET Framework)
- .NET Framework 4.7.2
- Platform Target architecture: x64
- Operating system: Windows10

**NuGet-packages installed:**

- CefSharp.WinForms  87.1.132
- CefSharp.Common    87.1.132
- CefSharp.OffScreen 87.1.132
- CefSharp.redist    x86.87.1.13
- CefSharp.redist    x64.87.1.13

**What I want to deploy:**

- Publish Wizard: From a CD-ROM or DVD-ROM

**Result:**
************** Exception Text **************
_System.IO.FileNotFoundException: Could not load file or assembly '**CefSharp.Core.Runtime.dll**' or one of its dependencies. The specified module could not be found.
File name: 'CefSharp.Core.Runtime.dll'
   at CefSharp.CefSettingsBase..ctor()
   at CefSharp.OffScreen.CefSettings..ctor()
   at CEFSimplified.Form1.Form1_Load(Object sender, EventArgs e)
   at System.Windows.Forms.Form.OnLoad(EventArgs e)
   at System.Windows.Forms.Form.OnCreateControl()
   at System.Windows.Forms.Control.CreateControl(Boolean fIgnoreVisible)
   at System.Windows.Forms.Control.CreateControl()
   at System.Windows.Forms.Control.WmShowWindow(Message& m)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.Form.WmShowWindow(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(IntPtr hWnd, Int32 msg, IntPtr wparam, IntPtr lparam)_

.**dll-files involved, all found on the computer in question:**
- KERNEL32.dll
- MSVCP140.dll
- VCRUNTIME140.dll
- api-ms-win-crt-runtime-l1-1-0.dll
- api-ms-win-crt-string-l1-1-0.dll
- api-ms-win-crt-heap-l1-1-0.dll
- libcef.dll
- USER32.dll
- mscoree.dll

- Microsoft Visual C++ 2005,2008, 2010, 2012,2013, 2015-2019 installed

**Comment by amaitland** [comment]
> CefSharpBuildAction Content /CefSharpBuildAction
Results in: The element CefSharpBuildAction beneath element is unrecognized.

What exactly did you add and Where did you add it?

Are you targeting x86 or x64 or AnyCpu?

**Reply by mrasoftGithub** [reply]
Thanks for reacting. Regarding the Platform Target architecture, I targeted **x64**,  Like in the instruction videos I did this in Visual Studio: **Configuration Manager** > **Platform** > **New Platform**  >**x64**. And in my Visual Studio Project x64 is shown as active solution platform.

In my Visual Studio Project this is shown: 
![ConfigurationManager](https://user-images.githubusercontent.com/78913022/107699558-8a980b80-6cb6-11eb-894c-ce14ea1154bf.jpg)

 And this:
![ActiveSolutionPlatform](https://user-images.githubusercontent.com/78913022/107699983-20339b00-6cb7-11eb-9259-ba15212dc206.jpg)

**Reply by amaitland** [reply]
Please answer both questions. Thanks

**Reply by mrasoftGithub** [reply]
I am sorry, I now will try to give an answer to your first question. In the project file I tried to insert this:
`<CefSharpBuildAction>Content</CefSharpBuildAction>`

The project file:
![ProjectFile1](https://user-images.githubusercontent.com/78913022/107701671-81f50480-6cb9-11eb-9e99-c76da13978c7.JPG)

Trying to load the project:
![ProjectFile2](https://user-images.githubusercontent.com/78913022/107701708-8e795d00-6cb9-11eb-913c-30552f619e18.JPG)

It seems i am doing something wrong regarding the Platform Target architecture and project file? 

Thanks.

**Reply by amaitland** [reply]
Add the CefSharpBuildAction property to the first PropertyGroup in your project file. You cannot add a property directly.

**Reply by mrasoftGithub** [reply]
Just adding the property to the first PropertyGroup gave the same error.
`<CefSharpBuildAction>Content</CefSharpBuildAction>`

So I added the property to this PropertyGroup:
`<PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">`
`<CefSharpBuildAction>Content</CefSharpBuildAction>`

The error disappears but I now got this:
![Installation_msg1](https://user-images.githubusercontent.com/78913022/107736062-6ca3c880-6d01-11eb-882f-f7e553e9a849.JPG)

What I found in the error log:

+ Downloading file:
///C:/CEFSimplified/publish/Application Files/CEFSimplified_1_0_0_0/CefSharp.Core.Runtime.dll.deploy 
did not succeed.

+ The file 
**'C:\Users\MRA\AppData\Local\Temp\Deployment\XJBRJ1BE.O6O\E67WROJV.TOM\CefSharp.Core.Runtime.dll' already exists.**


The Project File:
![Projectfile-PropertyGroup](https://user-images.githubusercontent.com/78913022/107735999-4d0ca000-6d01-11eb-9add-1ba80c775691.JPG)

Part of the Error log:
`PLATFORM VERSION INFO
	Windows 			: 10.0.19041.0 (Win32NT)
	Common Language Runtime 	: 4.0.30319.42000
	System.Deployment.dll 		: 4.8.4270.0 built by: NET48REL1LAST_C
	clr.dll 			: 4.8.4300.0 built by: NET48REL1LAST_C
	dfdll.dll 			: 4.8.4270.0 built by: NET48REL1LAST_C
	dfshim.dll 			: 10.0.19041.1 (WinBuild.160101.0800)

SOURCES
	Deployment url			: file:///C:/CEFSimplified/publish/CEFSimplified.application
	Application url			: file:///C:/CEFSimplified/publish/Application%20Files/CEFSimplified_1_0_0_0/CEFSimplified.exe.manifest

IDENTITIES
	Deployment Identity		: CEFSimplified.application, Version=1.0.0.0, Culture=neutral, PublicKeyToken=58de43f307f5cb82, processorArchitecture=amd64
	Application Identity		: CEFSimplified.exe, Version=1.0.0.0, Culture=neutral, PublicKeyToken=58de43f307f5cb82, processorArchitecture=amd64, type=win32

APPLICATION SUMMARY
	* Installable application.

ERROR SUMMARY
	Below is a summary of the errors, details of these errors are listed later in the log.
	* Activation of C:\CEFSimplified\publish\CEFSimplified.application resulted in exception. Following failure messages were detected:
	* 
		+ Downloading file:
///C:/CEFSimplified/publish/Application Files/CEFSimplified_1_0_0_0/CefSharp.Core.Runtime.dll.deploy 
did not succeed.
		+ The file 
**'C:\Users\MRA\AppData\Local\Temp\Deployment\XJBRJ1BE.O6O\E67WROJV.TOM\CefSharp.Core.Runtime.dll' already exists.**`

**Reply by amaitland** [reply]
> Just adding the property to the first PropertyGroup gave the same error.

If you see exactly the same error then it suggests you made a mistake. 

Adding the property to only the debug config is a problem if you publish a release version. 

Also prefer32bit shouldn't be set for x64 target.

**Reply by mrasoftGithub** [reply]
My .csproj projectfile has 9 property groups (for simplicity i address them with a number). To which property group do I have to add this property? Will one do? Or do I have to add the property to all property groups depending on the Platform Target architecture (x64, x86, AnyCPU) and the build (debug , release)?

`<CefSharpBuildAction>Content</CefSharpBuildAction>`

So far my attemps applied to adding the property to PropertyGroup 1 and 4 and I did a debug-build with a x64 target architecture.

It seems I did something wrong the first time but now I got this error in both cases:

![Installation_msg1](https://user-images.githubusercontent.com/78913022/107845230-4ef66200-6dda-11eb-9b50-4ef7f165e0ea.JPG)

With this message in the error log:
_'C:\Users\MRA\AppData\Local\Temp\Deployment\XJBRJ1BE.O6O\E67WROJV.TOM\CefSharp.Core.Runtime.dll' already exists._

I also changed the Prefer32Bit property to false but the results were the same (in the print screens the value is set to  true)
`<Prefer32Bit>false</Prefer32Bit>`

Property group 1) applies to AnyCPU?
`<PropertyGroup>`
`<Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>`
`<Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>`

![Projectfile-PropertyGroup_1](https://user-images.githubusercontent.com/78913022/107845184-ef985200-6dd9-11eb-86df-a97ef7552c77.JPG)

Property group 2)
`<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">`
`<PlatformTarget>AnyCPU</PlatformTarget>`

Property group 3)
`<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">`
`<PlatformTarget>AnyCPU</PlatformTarget>`

Property group 4)
`<PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x64'">`
`<DebugSymbols>true</DebugSymbols>`
`<OutputPath>bin\x64\Debug\</OutputPath>`

![Projectfile-PropertyGroup_4](https://user-images.githubusercontent.com/78913022/107845191-faeb7d80-6dd9-11eb-8772-5b2a8b231326.JPG)

Property group 5)
`<PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Release|x64'">`
`<OutputPath>bin\x64\Release\</OutputPath>`

Property group 6)
`<PropertyGroup>`
`<ManifestCertificateThumbprint>EB1E...`

Property group 7)
`<PropertyGroup>`
`<ManifestKeyFile>CEFSimplified_Temp`

Property group 8)
`<PropertyGroup>`
`<GenerateManifests>true</GenerateManifests>`

Property group 9)
`<PropertyGroup>`
`<SignManifests>true</SignManifests>`

**Reply by amaitland** [reply]
Which platform did you publish for? Publishing for AnyCpu is not currently supported with CefSharpBuildAction set to Content. 

I've never seen the error before,  if you are targeting AnyCPU when you publish then it's likely the two arch specific implementations of CefSharp.Core.Runtime.dll that's causing the problem. 

Currently targetplatform of x86 and x64 are supported.

**Reply by mrasoftGithub** [reply]
My intention is to publish for a x64 platform:
![ActiveSolutionPlatform](https://user-images.githubusercontent.com/78913022/107846736-42780680-6de6-11eb-9ce6-27bc18bc6b4f.jpg)

And on looking at my Visual Studio 2017-screen I should say Visual Studio has has it set to a x64 platform:
![ConfigurationManager](https://user-images.githubusercontent.com/78913022/107846764-8408b180-6de6-11eb-8b85-2025bd83d34d.jpg)

Furthermore I did create that project "from scratch". It is not cloned from an existing project from a GitHub-repository. I did all several times on different (independent from each other) computers and I encountered the same errors (or on all computers I did something consistently wrong but what??). The project itself only consists of one form with some CEFSharp NuGet-packages and C# code. It is even more simplified than the CEFSharp example project found on GitHub. 

I can put my project in a public GitHub repository if you are willing to have a look  at my project and if you have the opportunity and time to do that. Needless to say I  certainly do appreciate that you have paid attention to this discussion so far. 

Best Regards,

**Reply by amaitland** [reply]
Can you test with `VS2019`? I suspect the problem only reproduces with `VS2017`.

I'm testing with `VS2019` which I believe is quite a bit different in terms of `ClickOnce`.

**Reply by mrasoftGithub** [reply]
I did the whole exercise again from scratch with Visual Studio **2019** (Community Version **16.8.5**) and the error _**did not occur and it was installed without problems**_.

**This came  to my mind**:
1) By default cef.redist.x64.87.1.13 and cef.redist.x.86.87.1.13 are installed (I presume as a dependency package):
![011](https://user-images.githubusercontent.com/78913022/107874677-55a8d600-6ebb-11eb-9d90-9b409d1ec6f9.JPG)

2) I decided the install the new versions:
![015](https://user-images.githubusercontent.com/78913022/107874691-7c670c80-6ebb-11eb-8059-b55a60a83996.JPG)

3) and this as well:
![017](https://user-images.githubusercontent.com/78913022/107874853-d7e5ca00-6ebc-11eb-8c2f-de4e65fb2ea0.JPG)

4) but its build gave errors:
![020](https://user-images.githubusercontent.com/78913022/107874720-bdf7b780-6ebb-11eb-8524-a917438dadc8.JPG)

5) so I uninstalled the packages in question and i did proceed with the NuGet-packages below:
![021](https://user-images.githubusercontent.com/78913022/107874728-d5cf3b80-6ebb-11eb-824d-ac408880f7a8.JPG)

**With the two packages installed I did this**:

1) The x64 project platform:
![006](https://user-images.githubusercontent.com/78913022/107874527-8dfbe480-6eba-11eb-9384-7c2af8410fae.JPG)

2) Adding this property to the .csproj  project file:
`<CefSharpBuildAction>Content</CefSharpBuildAction>`
![022](https://user-images.githubusercontent.com/78913022/107874571-c8fe1800-6eba-11eb-84e4-93ad35db7afa.JPG)

3) Publishing
![023](https://user-images.githubusercontent.com/78913022/107874582-db785180-6eba-11eb-89e8-698b42ec1551.JPG)

4) Successfully installed
![031](https://user-images.githubusercontent.com/78913022/107874604-f945b680-6eba-11eb-80fe-15f03e1bfb3f.JPG)

**Reply by amaitland** [reply]
Thanks for confirming. The newer versions of cef.redist are for the 88 pre release. 

A quick search and it looks suggests VS 2017 has likely generated a manifest with a duplicate entry. We can probably hack around this.

https://stackoverflow.com/questions/6793777/clickonce-file-already-exists-error-why-is-a-dll-file-trying-to-be-copied-tw

**Reply by amaitland** [reply]
For `VS2017` you should be able to add the following to your project file to workaround the issue.

```xml
<Target Name="CefSharpClickOnceVS2017" BeforeTargets="GenerateApplicationManifest">
	<ItemGroup>
		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.dll" />
		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.pdb" />
		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.xml" />
	</ItemGroup>
</Target>
```

The duplicate reference should be removed from the manifest.

**Reply by mrasoftGithub** [reply]
I added your `<Target Name="CefSharpClickOnceVS2017" BeforeTargets="GenerateApplicationManifest">` ...`</Target>` to the project file and the workaround **did work**. Thanks.

1) Visual Studio 2017
![DoesWork00](https://user-images.githubusercontent.com/78913022/107980014-d81abe00-6fbf-11eb-90c7-bb8c638cdb60.JPG)

2) I did include this as well: `<CefSharpBuildAction>Content</CefSharpBuildAction>`
![DoesWork01](https://user-images.githubusercontent.com/78913022/107980120-00a2b800-6fc0-11eb-9b9f-b90886765b2a.JPG)

3) The bypass: `<Target Name="CefSharpClickOnceVS2017" BeforeTargets="GenerateApplicationManifest">` ...`</Target>`
![DoesWork02](https://user-images.githubusercontent.com/78913022/107980204-26c85800-6fc0-11eb-927d-4ae6f9224a20.JPG)

4) Successfully installed:
![DoesWork03](https://user-images.githubusercontent.com/78913022/107980258-39db2800-6fc0-11eb-923c-9d126dd9992e.JPG)

**Comment by mrasoftGithub** [comment]
I am sorry but just adding this to a PropertyGroup in Visual Studio 2017 did not seem to work:
`<CefSharpBuildAction>Content</CefSharpBuildAction>`

**Adding this to the project file solved the problem in which duplicates are removed from the manifest file**:

`<Target Name="CefSharpClickOnceVS2017" BeforeTargets="GenerateApplicationManifest">`
`	<ItemGroup>`
`		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.dll" />`
`		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.pdb" />`
`		<_DeploymentManifestFiles Remove="$(SolutionDir)packages\CefSharp.Common.87.1.132\buildTransitive\..\CefSharp\$(PlatformTarget)\CefSharp.Core.Runtime.xml" />`
`	</ItemGroup>`
`</Target>`

**Or adding this to the project file helped as well**:

`<ItemGroup>`
`    <Content Include="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\**\*" Exclude="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\x64\**\*;$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\locales\**\*.pak">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`      <Visible>false</Visible>`
`    </Content>`
`    <Content Include="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\**\en-US.*">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`    </Content>`
`    <Content Include="$(SolutionDir)packages\cef.redist.x64.87.1.13\CEF\x64\**\*">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`      <Visible>false</Visible>`
`    </Content>`
`    <Content Include="$(SolutionDir)packages\CefSharp.Common.87.1.132\CefSharp\x64\**\CefSharp.BrowserSubprocess.*">`
`      <Link>%(RecursiveDir)%(Filename)%(Extension)</Link>`
`      <Visible>false</Visible>`
`    </Content>`
`</ItemGroup>`

It seems **v87.1.1.132** still needs some things to be added to the project file in a ClickOnce Visual Studio 2017-project. 

Thanks for responding to my discussion.

Best Regards,
